xbmc/addons/gui/GUIDialogAddonInfo.cpp

```diff
@@ -303,28 +303,31 @@ void CGUIDialogAddonInfo::OnUpdate()
     if (i != -1)
     {
       Close();
-      // turn auto updating off if downgrading
-
-      if (m_localAddon->Version() > versions[i].first)
-        CServiceBroker::GetAddonMgr().AddUpdateRuleToList(m_localAddon->ID(),
-                                                          AddonUpdateRule::PIN_OLD_VERSION);
-
-      // turn auto updating on if upgrading or changing origin
-      // (but not when changing to local cache)
-
-      if ((m_localAddon->Version() < versions[i].first &&
-           m_localAddon->Origin() == versions[i].second) ||
-          (m_localAddon->Origin() != versions[i].second && versions[i].second != LOCAL_CACHE))
-        CServiceBroker::GetAddonMgr().RemoveUpdateRuleFromList(m_localAddon->ID(),
-                                                               AddonUpdateRule::PIN_OLD_VERSION);
 
       if (versions[i].second == LOCAL_CACHE)
         CAddonInstaller::GetInstance().InstallFromZip(
             StringUtils::Format("special://home/addons/packages/%s-%s.zip",
                                 m_localAddon->ID().c_str(), versions[i].first.asString().c_str()));
-      else
-        CAddonInstaller::GetInstance().Install(m_localAddon->ID(), versions[i].first,
-                                               versions[i].second);
+      else if (CAddonInstaller::GetInstance().Install(m_localAddon->ID(), versions[i].first,
+                                                      versions[i].second))
+      {
+        // turn auto updating * OFF *
+        // if downgrading within the same origin
+
+        if (m_localAddon->Origin() == versions[i].second &&
+            m_localAddon->Version() > versions[i].first)
+          CServiceBroker::GetAddonMgr().AddUpdateRuleToList(m_localAddon->ID(),
+                                                            AddonUpdateRule::PIN_OLD_VERSION);
+
+        // turn auto updating * ON *
+        // if changing to a different origin or upgrading within the same origin
+
+        if (m_localAddon->Origin() != versions[i].second ||
+            (m_localAddon->Version() < versions[i].first &&
+             m_localAddon->Origin() == versions[i].second))
+          CServiceBroker::GetAddonMgr().RemoveUpdateRuleFromList(m_localAddon->ID(),
```

Comment:
Does this logic test for upgrading to latest of only a version later than the current version?

Reply:
only a version later than the current.
means when we have versions `1`, `2`, and `3` available..
..and we upgrade from version `1` to version `2` -> we will remove the pin.

IMO not 100% perfect and we should unpin only when upgrading to the latest.
came across this thought some time ago...

Reply:
Agreed, it should only unpin when upgraded to latest.